libnssfix fails to build under Ubuntu 22.04

Noted here:
https://github.com/JonathonReinhart/staticx/pull/194#issuecomment-1624289769

```
gcc -o scons_build/debug/libnssfix/libnssfix.so -Wl,--fatal-warnings -Wl,--no-as-needed -nostdlib -Wl,--no-undefined -shared scons_build/debug/libnssfix/nssfix.os -Lscons_build/debug/lib scons_build/debug/libnssfix/libc.so.6 -lnss_dns -lnss_files
/bin/ld: cannot find -lnss_dns: No such file or directory
/bin/ld: cannot find -lnss_files: No such file or directory
```

--- a/.github/workflows/build-test.yml
+++ b/.github/workflows/build-test.yml
@@ -7,12 +7,26 @@ on:
 
 jobs:
   build:
-
-    runs-on: ubuntu-22.04  # TODO: would be nice to matrix this w/ 20.04
     strategy:
       matrix:
-        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
+        # It's undocumented, but matrix variable arrays can be maps, too!
+        # https://stackoverflow.com/a/68940067/119527
+        os:
+          - version: "ubuntu-20.04"
+            native-python: "3.8"
+          - version: "ubuntu-22.04"
+            native-python: "3.10"
+        python-version:
+          - "3.7"
+          - "3.8"
+          - "3.9"
+          - "3.10"
+          - "3.11"
       fail-fast: False
+
+    name: "${{matrix.os.version}}: Python ${{matrix.python-version}}"
+
+    runs-on: ${{ matrix.os.version }}
 
     steps:
     - uses: actions/checkout@v3
@@ -25,7 +39,7 @@ jobs:
     # with StaticX. See #224.
     - name: Set up Python ${{ matrix.python-version }}
       uses: deadsnakes/action@v3.0.1
-      if: "(matrix.python-version != '3.10')"  # Already installed in ubuntu-22.04
+      if: "(matrix.python-version != matrix.os.native-python)"  # Already installed
       with:
         python-version: ${{ matrix.python-version }}
 

--- a/.github/workflows/build-test.yml
+++ b/.github/workflows/build-test.yml
@@ -8,7 +8,7 @@ on:
 jobs:
   build:
 
-    runs-on: ubuntu-20.04  # See #245
+    runs-on: ubuntu-22.04  # TODO: would be nice to matrix this w/ 20.04
     strategy:
       matrix:
         python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
@@ -25,7 +25,7 @@ jobs:
     # with StaticX. See #224.
     - name: Set up Python ${{ matrix.python-version }}
       uses: deadsnakes/action@v3.0.1
-      if: "(matrix.python-version != '3.8')"  # Already installed in ubuntu-20.04
+      if: "(matrix.python-version != '3.10')"  # Already installed in ubuntu-22.04
       with:
         python-version: ${{ matrix.python-version }}
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,9 @@ All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](https://semver.org/).
 
 ## [Unreleased]
+### Fixed
+- Fixed issue causing libnssfix link failure when building on GLIBC 2.34 ([#255])
+
 ### Removed
 - Dropped support for Python 3.5 and 3.6 ([#247])
 
@@ -334,3 +337,4 @@ Initial release
 [#237]: https://github.com/JonathonReinhart/staticx/pull/237
 [#238]: https://github.com/JonathonReinhart/staticx/pull/238
 [#247]: https://github.com/JonathonReinhart/staticx/pull/247
+[#255]: https://github.com/JonathonReinhart/staticx/pull/255

--- a/SConstruct
+++ b/SConstruct
@@ -66,6 +66,18 @@ conf = base_env.Configure(custom_tests=custom_tests)
 # Does our default libc have NSS?
 conf.env["HAS_NSS"] = conf.CheckNSS()
 
+# Starting in GLIBC 2.34, some NSS modules (files, dns) can be built-in to
+# libc.so. Determine which ones are builtin. See #245.
+conf.env["NSS_MODS_BUILTIN"] = []
+if conf.env["HAS_NSS"]:
+    maybe_builtin_nss_modules = ["files", "dns"]
+    for mod in maybe_builtin_nss_modules:
+        libname = "libnss_" + mod
+        # It the library doesn't exist, assume it is built-in.
+        if not conf.BasicCheckLib(libname):
+            conf.env.Append(NSS_MODS_BUILTIN=libname)
+    print("Builtin NSS modules:", conf.env["NSS_MODS_BUILTIN"])
+
 base_env = conf.Finish()
 
 ################################################################################

--- a/SConstruct
+++ b/SConstruct
@@ -59,10 +59,14 @@ def BuildSubdir(env, dirname):
 base_env.AddMethod(BuildSubdir)
 
 
+###########
+# Configure
 conf = base_env.Configure(custom_tests=custom_tests)
-has_nss = conf.CheckNSS()
-conf.Finish()
 
+# Does our default libc have NSS?
+conf.env["HAS_NSS"] = conf.CheckNSS()
+
+base_env = conf.Finish()
 
 ################################################################################
 # Bootloader
@@ -79,7 +83,7 @@ for env in bootloader_env.ModeEnvs():
 
 ################################################################################
 # nssfix
-if has_nss:
+if env['HAS_NSS']:
     for env in base_env.ModeEnvs():
         env.InstallAs('#staticx/assets/$MODE/libnssfix.so', env.BuildSubdir('libnssfix'))
 else:

--- a/libnssfix/SConscript
+++ b/libnssfix/SConscript
@@ -22,6 +22,16 @@ env.NsswitchConfH(
     source = nsswitch_conf,
 )
 
+
+# Link against the configured NSS module libraries.
+# These aren't needed directly for the library itself, but this ensures
+# that they will be available for nss at runtime (via dlopen), and
+# allows their paths to be easily discovered via ldd(1).
+nss_libs = set(env.GetNsswitchLibs(nsswitch_conf))
+
+# Exclude those already built-in to glibc.
+nss_libs -= set(env["NSS_MODS_BUILTIN"])
+
 # Build libnssfix.so
 libnssfix = env.SharedLibrary(
     target = 'nssfix',
@@ -50,12 +60,7 @@ libnssfix = env.SharedLibrary(
     ],
     LIBS = [
         libc_stub,
-
-        # These aren't needed directly for the library itself, but this ensures
-        # that they will be available for nss at runtime (via dlopen), and
-        # allows their paths to be easily discovered via ldd(1).
-        env.GetNsswitchLibs(nsswitch_conf),
-    ],
+    ] + list(nss_libs),
 )
 
 Return('libnssfix')

--- a/site_scons/conftest.py
+++ b/site_scons/conftest.py
@@ -14,6 +14,26 @@ int main(void)
     return ok
 
 
+def BasicCheckLib(context, libname):
+    """Check for a C library.
+
+    Similar to built-in CheckLib but works around
+    https://github.com/SCons/scons/issues/4373.
+    """
+    context.Message("Checking for library %s... " % libname)
+
+    oldLIBS = context.AppendLIBS([libname])
+    ok = context.TryLink(
+        "int main(void) { return 0; }",
+        ".c"
+    )
+    context.SetLIBS(oldLIBS)
+
+    context.Result(ok)
+    return ok
+
+
 custom_tests = dict(
     CheckNSS = CheckNSS,
+    BasicCheckLib = BasicCheckLib,
 )

